extensions/api/data-management/transferprocess/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/transferprocess/service/TransferProcessServiceImplTest.java

```diff
@@ -0,0 +1,231 @@
+
+        // Act
+        Result<?> result = service.cancel(id);
+
+        // Assert
+        assertThat(result.failed()).isTrue();
+        assertThat(result.getFailureMessages()).containsExactly("Cannot cancel a transfer process in state " + state);
+        verifyNoInteractions(manager);
+    }
+
+    @Test
+    void cancel_whenNotFound() {
+        // Act
+        Result<?> result = service.cancel(id);
+
+        // Assert
+        assertThat(result.failed()).isTrue();
+        assertThat(result.getFailureMessages()).containsExactly("Not found " + id);
+    }
+
+    public static List<TransferProcessStates> cancellableStates() {
+        var states = new ArrayList<>(Arrays.asList(TransferProcessStates.values()));
+        states.removeAll(nonCancellableStates());
+        return states;
+    }
+
+    public static List<TransferProcessStates> nonCancellableStates() {
+        return List.of(
+                COMPLETED,
+                ENDED,
+                ERROR
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("deprovisionableStates")
+    void deprovision(TransferProcessStates state) {
+        // Arrange
+        var process = transferProcess(state);
+        when(store.find(id)).thenReturn(process);
+
+        // Act
+        var result = service.deprovision(id);
+
+        // Assert
+        assertThat(result.succeeded()).isTrue();
+        verify(manager).enqueueCommand(commandCaptor.capture());
+        assertThat(commandCaptor.getValue()).isInstanceOf(DeprovisionRequest.class);
+        assertThat(commandCaptor.getValue().getTransferProcessId())
+                .isEqualTo(id);
+    }
+
+    @ParameterizedTest
+    @MethodSource("nonDeprovisionableStates")
+    void deprovision_whenNonDeprovisionable(TransferProcessStates state) {
+        // Arrange
+        var process = transferProcess(state);
+        when(store.find(id)).thenReturn(process);
+
+        // Act
+        Result<?> result = service.deprovision(id);
+
+        // Assert
+        assertThat(result.failed()).isTrue();
+        assertThat(result.getFailureMessages()).containsExactly("Cannot deprovision a transfer process in state " + state);
+        verifyNoInteractions(manager);
+    }
+
+    @Test
+    void deprovision_whenNotFound() {
+        // Act
+        Result<?> result = service.deprovision(id);
+
+        // Assert
+        assertThat(result.failed()).isTrue();
+        assertThat(result.getFailureMessages()).containsExactly("Not found " + id);
+    }
+
+    public static List<TransferProcessStates> nonDeprovisionableStates() {
+        var states = new ArrayList<>(Arrays.asList(TransferProcessStates.values()));
+        states.removeAll(deprovisionableStates());
+        return states;
+    }
+
+    public static List<TransferProcessStates> deprovisionableStates() {
+        return List.of(
+                COMPLETED,
+                DEPROVISIONING,
+                DEPROVISIONED,
+                ENDED,
```

Comment:
I am a bit confused by the values here. It seems that deprovisioning has no effect if the TP is in `ENDED` or `CANCELLED` states. Is still correct to have `TransferProcessServiceImpl` enqueue a command in this case?

Reply:
Per the story we should implement if the state transition is allowed, which is the case here. I'll flag this point in the upstream PR for confirmation.